internal/toolexec/proxy/proxy.go

```diff
@@ -0,0 +1,120 @@
+package proxy
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+type (
+	// CommandType represents a Go compilation command type
+	// Currently support types are:
+	// - compile
+	// - link
+	// - unknown
+	CommandType string
+	// Command represents a Go compilation command
+	Command interface {
+		Args() []string
+		ReplaceParam(param string, val string) error
+		Stage() string
+		Type() CommandType
+	}
+
+	// CommandProcessor is a function that takes a command as input
+	// and is allowed to modify it or read its data
+	CommandProcessor[T Command] func(T)
+
+	commandFlagSet struct {
+		Output string `ddflag:"-o"`
+	}
+
+	// command is the default unknown command type
+	// Can be used to compose specific Command implementations
+	command struct {
+		args []string
+		// paramPos is the index in args of the *value* provided for the parameter stored in the key
+		paramPos map[string]int
+		flags    commandFlagSet
+	}
+)
+
+const (
+	CommandTypeUnknown CommandType = "unknown"
+	CommandTypeCompile             = "compile"
+	CommandTypeLink                = "link"
```

Comment:
IIRC, you need to type these explicitly if you're not using `iota`:

```suggestion
	CommandTypeUnknown CommandType = "unknown"
	CommandTypeCompile CommandType = "compile"
	CommandTypeLink CommandType    = "link"
```

It might also actually be more efficient to use `int` based constants instead of strings, for example when performing `switch`.

Reply:
This was facilitating the transition from parsed string to command type representation.
I don't feel strongly about it though so I went ahead and used int instead, and added a small switch case to go from string to command type